### ISSUE TITLE
Unable to build nyxt on Debian 11 VM with SBCL 2.3.11/2.3.7

### ISSUE THREAD

**Issue by Kyuvi** [atlas-engineer/nyxt#3298]
**Describe the bug**
I am using a Debian 11 VM on QubesOS, I got to install nyxt through guix, but all attempts to build nyxt from the source code have failed.

**Precise recipe to reproduce the issue**
download  nyxt-3.10.0-source-with-submodules.tar.xz and unzip file
cd into created folder 
run make all

(I also can not build using ~guix through emacs or from~ the lisp repl, but I think I should open a separate issue for those but I can include them on this thread if that is better.)

**Information**
- OS name+version:  Debian 11 (VM on QubesOS)
  - Lisp implementation/version (if built from source): 2.3.11 and 2.3.7
 

**Output when started from a shell**
** NOTE: output from "make -B all" as it is shorter

/home/user/Downloads/sbcl/sbcl-2.3.11-install/bin/sbcl  --no-userinit --non-interactive --eval '(require "asdf")' --eval '(when (string= "true" "true") (setf asdf:*default-source-registries* (list (quote asdf/source-registry:environment-source-registry))) (asdf:clear-configuration) (asdf:load-asd "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules//libraries/nasdf/nasdf.asd") (asdf:load-asd "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules//nyxt.asd") (asdf:load-system :nyxt/submodules))' --eval '(asdf:load-asd "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules//libraries/nasdf/nasdf.asd")' --eval '(asdf:load-asd "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules//nyxt.asd")' --eval '(when (and (find-package :cffi) (uiop:getenv "GUIX_ENVIRONMENT")) (pushnew (pathname (format nil "~a/lib/" (uiop:getenv "GUIX_ENVIRONMENT"))) (symbol-value (read-from-string "cffi:*foreign-library-directories*" )) :test (quote equal)))' --eval '(asdf:load-system :nyxt/gi-gtk-application)' \
	--eval '(asdf:make :nyxt/gi-gtk-application)' \
	--eval '(uiop:quit)' || (printf "\n%s\n%s\n" "Compilation failed, see the above stacktrace." && exit 1)
This is SBCL 2.3.11, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
; running ("git" "-C"
           "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/"
           "submodule" "update" "--init" "--force" "--jobs" "4")
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
; CL_SOURCE_REGISTRY: "/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build//:/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/:"
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/trivial-garbage/trivial-garbage.asd" contains definition for system "trivial-garbage-tests". Please only define "trivial-garbage" and secondary systems with a name starting with "trivial-garbage/" (e.g. "trivial-garbage/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/calispel/calispel.asd" contains definition for system "calispel-test". Please only define "calispel" and secondary systems with a name starting with "calispel/" (e.g. "calispel/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/cl-ppcre/cl-ppcre.asd" contains definition for system "cl-ppcre-test". Please only define "cl-ppcre" and secondary systems with a name starting with "cl-ppcre/" (e.g. "cl-ppcre/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/flexi-streams/flexi-streams.asd" contains definition for system "flexi-streams-test". Please only define "flexi-streams" and secondary systems with a name starting with "flexi-streams/" (e.g. "flexi-streams/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/cl-ppcre/cl-ppcre-unicode.asd" contains definition for system "cl-ppcre-unicode-test". Please only define "cl-ppcre-unicode" and secondary systems with a name starting with "cl-ppcre-unicode/" (e.g. "cl-ppcre-unicode/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/s-xml/s-xml.asd" contains definition for system "s-xml.test". Please only define "s-xml" and secondary systems with a name starting with "s-xml/" (e.g. "s-xml/test") in that file.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/s-xml/s-xml.asd" contains definition for system "s-xml.examples". Please only define "s-xml" and secondary systems with a name starting with "s-xml/" (e.g. "s-xml/test") in that file.
WARNING: The CL-Containers system would enjoy having asdf-system-connections around. See
http://www.cliki.net/asdf-system-connections for details and download
instructions.
WARNING: System definition file #P"/home/user/Downloads/Nyxt/nyxt-3.10.0-source-with-submodules/_build/montezuma/montezuma.asd" contains definition for system "montezuma-tests". Please only define "montezuma" and secondary systems with a name starting with "montezuma/" (e.g. "montezuma/test") in that file.
;;; Computing Hangul syllable namesWARNING: redefining S-SERIALIZATION::SERIALIZE-SEXP-INTERNAL in DEFGENERIC
WARNING: redefining CL-PREVALENCE:GET-ID in DEFGENERIC
WARNING: redefining CL-PREVALENCE:GET-NAME in DEFGENERIC
WARNING: redefining (COMMON-LISP:SETF CL-PREVALENCE:GET-NAME) in DEFGENERIC
WARNING: redefining CL-PREVALENCE:GET-SIZE in DEFGENERIC
WARNING: redefining CL-PREVALENCE:GET-MIME-TYPE in DEFGENERIC
WARNING: redefining (COMMON-LISP:SETF CL-PREVALENCE:GET-MIME-TYPE) in DEFGENERIC
WARNING: redefining CL-PREVALENCE:GET-KEYWORDS in DEFGENERIC
WARNING: redefining (COMMON-LISP:SETF CL-PREVALENCE:GET-KEYWORDS) in DEFGENERIC
WARNING: redefining DISSECT:STACK in DEFUN
WARNING: redefining DISSECT:RESTARTS in DEFUN
WARNING: redefining DISSECT::STACK-CAPPER in DEFUN
WARNING: redefining DISSECT::STACK-TRUNCATOR in DEFUN
WARNING: redefining DISSECT:WITH-TRUNCATED-STACK in DEFMACRO
WARNING:
   redefining CONSUME-ITEM (#<BUILT-IN-CLASS COMMON-LISP:SYMBOL>
                            #<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>) in DEFMETHOD
WARNING: redefining LOG4CL-IMPL:APPENDER-ADDED in DEFGENERIC
WARNING: redefining LOG4CL-IMPL:APPENDER-REMOVED in DEFGENERIC
WARNING: redefining LOG4CL-IMPL::PROPERTY-ALIST in DEFGENERIC
WARNING: redefining JSON::READ-JSON-TOKEN in DEFUN
WARNING: redefining JSON::PEEK-JSON-TOKEN in DEFUN
WARNING:
   redefining ENCODE-TO-STRING (#<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>) in DEFMETHOD
WARNING: redefining SERAPEUM:RUN-HOOK in DEFGENERIC
STYLE-WARNING:
   Generic function FSET:ITERATOR clobbers an earlier FTYPE proclamation
   (FUNCTION (T &KEY &ALLOW-OTHER-KEYS) (VALUES FUNCTION &REST T)) for the same
   name with (FUNCTION (T &KEY &ALLOW-OTHER-KEYS) *).
WARNING:
   The new FTYPE proclamation for CLSS:MATCH-SELECTOR
     (FUNCTION (LIST (OR PLUMP-DOM:NODE VECTOR LIST) KEYWORD)
      (VALUES (AND (VECTOR PLUMP-DOM:NODE) (NOT SIMPLE-ARRAY))))
   does not match the old FTYPE proclamation
     (FUNCTION (LIST (OR LIST PLUMP-DOM:NODE VECTOR) &OPTIONAL KEYWORD)
      (VALUES (AND (VECTOR PLUMP-DOM:NODE) (NOT SIMPLE-ARRAY)) &REST T))
WARNING:
   redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>) in DEFMETHOD
WARNING: redefining SLYNK::SYMBOL-INDENTATION in DEFUN
WARNING: redefining SLYNK::MACRO-INDENTATION in DEFUN
WARNING:
   bare references to struct types are deprecated. Please use (:POINTER
                                                               (:STRUCT
                                                                SQLITE3)) or (:STRUCT
                                                                              SQLITE3) instead.
WARNING:
   bare references to struct types are deprecated. Please use (:POINTER
                                                               (:STRUCT
                                                                SQLITE3-STMT)) or (:STRUCT
                                                                                   SQLITE3-STMT) instead.
WARNING:
   redefining SB-GRAY:STREAM-LINE-LENGTH (#<STANDARD-CLASS SB-GRAY:FUNDAMENTAL-CHARACTER-INPUT-STREAM>) in DEFMETHOD
WARNING:
   redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>) in DEFMETHOD
WARNING:
   redefining THREAD-FOR-EVALUATION (#<STRUCTURE-CLASS SWANK::MULTITHREADED-CONNECTION>
                                     #<SB-MOP:EQL-SPECIALIZER :FIND-EXISTING>) in DEFMETHOD
WARNING: redefining SWANK::SYMBOL-INDENTATION in DEFUN
WARNING: redefining SWANK::MACRO-INDENTATION in DEFUN
WARNING: redefining NYXT:EQUALS in DEFGENERIC

; compiling file "NYXT:SOURCE;MANUAL.LISP.NEWEST" (written 11 DEC 2023 03:56:59 PM):
Heap exhausted during garbage collection: 0 bytes available, 32 requested.
        Immobile Object Counts
 Gen layout fdefn symbol   code  Boxed   Cons    Raw   Code  SmMix  Mixed  LgRaw LgCode  LgMix Waste%       Alloc        Trig   Dirty GCs Mem-age
  1      0      0      0      0   5740    346   9803      8     24     52      0      0      0    7.3   484998800   280703274   15973   1  1.3350
  2   2865  41601   7437  31090   4948    731  10011     47    139    162     62      0    137    7.3   492984512   126056058   13911   1  0.2339
  3      0      0      0      0      0      0      0      0      0      0      0      0      0    0.0           0     2000000       0   0  0.0000
  4      0      0      0      0      0      0      0      0      0      0      0      0      0    0.0           0     2000000       0   0  0.0000
  5      0      0      0      0      0      0      0      0      0      0      0      0      0    0.0           0     2000000       0   0  0.0000
  6    763  16519  26487  18745    339    112      7      3      8     15      0      0     63    2.8    17429088     2000000      31   0  0.0000
Tot   3628  58120  33924  49835  11027   1189  19821     58    171    229     62      0    200    7.3   995410128 [92.7% of 1073741824 max]
GC control variables:
   *GC-INHIBIT* = true
   *GC-PENDING* = true
   *STOP-FOR-GC-PENDING* = false
Collection trigger variables:
   dynamic_space_size = 1073741824
   bytes_allocated = 995410128
   auto_gc_trigger = 617786323
   bytes_consed_between_gcs = 53687091
fatal error encountered in SBCL pid 13433 tid 13433:
Heap exhausted, game over.

   0: fp=0x7fb49099ddc8 pc=0x52de0c2f SB-REGALLOC::GROW-SC
   1: fp=0x7fb49099de78 pc=0x52b538bb SB-REGALLOC::PACK-TN
   2: fp=0x7fb49099df28 pc=0x52b53e60 (FLET SB-REGALLOC::PACK-TNS :IN SB-REGALLOC::PACK)
   3: fp=0x7fb49099dfd8 pc=0x52b52f9d SB-REGALLOC::PACK
   4: fp=0x7fb49099e040 pc=0x52d465a0 SB-C::%COMPILE-COMPONENT
   5: fp=0x7fb49099e060 pc=0x52b3b008 SB-C::COMPILE-COMPONENT
   6: fp=0x7fb49099e0a8 pc=0x52b8a080 SB-C::COMPILE-TOPLEVEL
   7: fp=0x7fb49099e0d0 pc=0x52c10a06 SB-C::CONVERT-AND-MAYBE-COMPILE
   8: fp=0x7fb49099e148 pc=0x52c112ee SB-C::PROCESS-TOPLEVEL-FORM
   9: fp=0x7fb49099e178 pc=0x52c10bf8 SB-C::PROCESS-TOPLEVEL-PROGN
  10: fp=0x7fb49099e1f0 pc=0x52c11261 SB-C::PROCESS-TOPLEVEL-FORM
  11: fp=0x7fb49099e268 pc=0x52c113d6 SB-C::PROCESS-TOPLEVEL-FORM
  12: fp=0x7fb49099e338 pc=0x52d4c7ba (LAMBDA (SB-KERNEL::FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN SB-C::SUB-COMPILE-FILE)
  13: fp=0x7fb49099e3e8 pc=0x52c103bb SB-C::%DO-FORMS-FROM-INFO
  14: fp=0x7fb49099e4a8 pc=0x52d4c319 (LAMBDA () :IN SB-C::SUB-COMPILE-FILE)
  15: fp=0x7fb49099e550 pc=0x52c0efcb (FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT)
  16: fp=0x7fb49099e6a0 pc=0x52d4bba9 SB-C::SUB-COMPILE-FILE
  17: fp=0x7fb49099e7e0 pc=0x5310a3b9 COMPILE-FILE
  18: fp=0x7fb49099e868 pc=0x5356d3cb UIOP/UTILITY::CALL-WITH-MUFFLED-CONDITIONS
  19: fp=0x7fb49099e898 pc=0x53577f35 UIOP/PATHNAME::CALL-WITH-ENOUGH-PATHNAME
  20: fp=0x7fb49099e980 pc=0x535923e9 UIOP/LISP-BUILD::COMPILE-FILE*
  21: fp=0x7fb49099e9d8 pc=0x53658914 NASDF::FAIL-ON-WARNINGS
  22: fp=0x7fb49099ea28 pc=0x535ba9ab ASDF/LISP-ACTION::PERFORM-LISP-COMPILATION
  23: fp=0x7fb49099ea70 pc=0x535f2b41 (SB-PCL::EMF ASDF/ACTION::PERFORM)
  24: fp=0x7fb49099eaf0 pc=0x535b2531 (LAMBDA () :IN ASDF/ACTION::CALL-WHILE-VISITING-ACTION)
  25: fp=0x7fb49099ecd8 pc=0x535b8efe (SB-PCL::FAST-METHOD ASDF/ACTION::PERFORM-WITH-RESTARTS :AROUND (T T))
  26: fp=0x7fb49099ed10 pc=0x535ca825 (SB-PCL::FAST-METHOD ASDF/PLAN::PERFORM-PLAN (T))
  27: fp=0x7fb49099edf0 pc=0x52c0ee1d (FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT)
  28: fp=0x7fb49099ee68 pc=0x535ca2c9 (SB-PCL::FAST-METHOD ASDF/PLAN::PERFORM-PLAN :AROUND (T))
  29: fp=0x7fb49099eeb0 pc=0x535cc704 (SB-PCL::FAST-METHOD ASDF/OPERATE::OPERATE (ASDF/OPERATION::OPERATION ASDF/COMPONENT::COMPONENT))
  30: fp=0x7fb49099ef08 pc=0x535ef81e (SB-PCL::EMF ASDF/OPERATE::OPERATE)
  31: fp=0x7fb49099efd0 pc=0x535cbb76 (LAMBDA () :IN ASDF/OPERATE::OPERATE)
  32: fp=0x7fb49099f098 pc=0x535cb6e2 (SB-PCL::FAST-METHOD ASDF/OPERATE::OPERATE :AROUND (T T))
  33: fp=0x7fb49099f0f0 pc=0x535ef81e (SB-PCL::EMF ASDF/OPERATE::OPERATE)
  34: fp=0x7fb49099f1b8 pc=0x535cbb76 (LAMBDA () :IN ASDF/OPERATE::OPERATE)
  35: fp=0x7fb49099f280 pc=0x535cb6e2 (SB-PCL::FAST-METHOD ASDF/OPERATE::OPERATE :AROUND (T T))
  36: fp=0x7fb49099f4a8 pc=0x535a7714 ASDF/SESSION::CALL-WITH-ASDF-SESSION
  37: fp=0x7fb49099f570 pc=0x535cbfa9 (LAMBDA () :IN ASDF/OPERATE::OPERATE)
  38: fp=0x7fb49099f798 pc=0x535a7714 ASDF/SESSION::CALL-WITH-ASDF-SESSION
  39: fp=0x7fb49099f860 pc=0x535cb6e2 (SB-PCL::FAST-METHOD ASDF/OPERATE::OPERATE :AROUND (T T))
  40: fp=0x7fb49099f890 pc=0x535cdc62 ASDF/OPERATE::LOAD-SYSTEM
  41: fp=0x7fb49099f968 pc=0x52ba601b SB-INT::SIMPLE-EVAL-IN-LEXENV
  42: fp=0x7fb49099f980 pc=0x52a516f2 EVAL
  43: fp=0x7fb49099fb90 pc=0x52e0688a SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS
  44: fp=0x7fb49099fdd8 pc=0x52c56ed7 SB-IMPL::TOPLEVEL-INIT
  45: fp=0x7fb49099fe70 pc=0x52f575b4 (FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP)
  46: fp=0x7fb49099ff30 pc=0x52f573bb (FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP)
  47: fp=0x7fb49099ffc8 pc=0x52f571e8 SB-IMPL::%START-LISP

Compilation failed, see the above stacktrace.

make: *** [makefile:51: nyxt] Error 1

**Comment by Kyuvi** [comment]
I should also note that this line

dynamic_space_size = 1073741824

does not change regardless of the amount of I put in the makefile or the amount of memory (up to 10gb) allocated to the VM

**Comment by Kyuvi** [comment]
>Out of curiosity, why are you trying to build it on a VM?

I use [QubesOS](www.qubes-os.org), It uses isolated VMs to hopefully maximize security,

Should I put the the logs of the other failed builds here or open a new issue?

The strangest being that when I try to build from a common lisp repl, it fails to find the  "CL-CFFI-GTK" package (with both ECL and Clisp as well as SBCL).

**Comment by Kyuvi** [comment]
I managed to find and solve the issues I was having, 

It was not to directly to do with the VM It had to do with the fact that the debian 11 repositories only contain SBCL up to 2.1.1 and that version created an unusable binary like in this issue [#3222](https://github.com/atlas-engineer/nyxt/issues/3222),  so I had installed sbcl 2.3.11 in its own directory and edited the 'LISP' variable in the makefile from "sbcl" to  "/home/user/Downloads/sbcl/sbcl-2.3.11-install/bin/sbcl".

the makefile then did not recognize it as "sbcl" and therefore did not allocate the dynamic-space-size. 

--- 
I also solved the issue with the build failing from the common lisp repl, I think there might be something missing from the documentation, because the makefile has the lines...

`(require "asdf")`
`(asdf:load-system :nyxt/submodules)`

which I found to be necessary before 
`(asdf:make :nyxt/gtk-application)`
 to get the compilation to work (at least for me ) along with requiring sbcl to be started with `--dynamic-space-size 3072`. This is not stated or made clear in the ["using lisp"](https://github.com/atlas-engineer/nyxt/blob/master/documents/README.org#using-lisp) part of the developer's documentation. 
While requiring asdf might be obvious, the next line is not and neither is setting the dynamic-space-size, unless one reads the makefile. 

I still could not get it to work with ECL(with this complaining about PARSE-BODY missing in UIOP:DRIVER)  and Clisp (due to what seems to be a memory space issue) though.

I will close this as all my issues have been solved.

Thanks
